File: day5/exercises-solutions/ping-pong.f90
PROGRAM pingpong
  USE mpi
  IMPLICIT NONE
  INTEGER :: size, rank, ierr, mesg1, mesg2, status(MPI_STATUS_SIZE)

  CALL MPI_INIT(ierr)
  CALL MPI_COMM_SIZE(MPI_COMM_WORLD,size,ierr)
  CALL MPI_COMM_RANK(MPI_COMM_WORLD,rank,ierr)

  IF (size /= 2) THEN
      IF (rank == 0) PRINT*,'must have exacrtly 2 MPI ranks'
      CALL MPI_ABORT(MPI_COMM_WORLD,1,ierr)
      STOP 'ABORT'
  END IF

  mesg1 = 42
  mesg2 = 0

  IF (rank == 0) THEN
      CALL MPI_SEND(mesg1, 1, MPI_INTEGER, 1, 0, MPI_COMM_WORLD, ierr)
      CALL MPI_RECV(mesg2, 1, MPI_INTEGER, 1, 0, MPI_COMM_WORLD, status, ierr)
      PRINT*,"Final result is", mesg2
  ELSE
      CALL MPI_RECV(mesg2, 1, MPI_INTEGER, 0, 0, MPI_COMM_WORLD, status, ierr)
      mesg1 = mesg2 + 99
      CALL MPI_SEND(mesg1, 1, MPI_INTEGER, 0, 0, MPI_COMM_WORLD, ierr)
  END IF

  CALL MPI_FINALIZE(ierr)
END PROGRAM pingpong

File: day5/exercises-solutions/ping-pong.c
#include <mpi.h>
#include <stdio.h>

int main(int argc, char **argv)
{
    int size,rank,mesg1,mesg2;
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD,&size);
    MPI_Comm_rank(MPI_COMM_WORLD,&rank);
    if (size != 2) {
        if (rank == 0) puts("must have exactly 2 MPI ranks");
        MPI_Abort(MPI_COMM_WORLD,1);
    }

    mesg1 = 42;
    mesg2 = 0;

    if (rank == 0) {
        MPI_Send(&mesg1, 1, MPI_INT, 1, 0, MPI_COMM_WORLD);
        MPI_Recv(&mesg2, 1, MPI_INT, 1, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
        printf("Final result is %d\n", mesg2);
    } else { 
        MPI_Recv(&mesg2, 1, MPI_INT, 0, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
        mesg1 = mesg2 + 99;
        MPI_Send(&mesg1, 1, MPI_INT, 0, 0, MPI_COMM_WORLD);
    }

    MPI_Finalize();
    return 0;
}
